### PR TITLE
Single source of truth

### PR DESCRIPTION
**Pull request by naterush**
# Description

This is an experimental PR that moves the `execute` function in the step performer so that it uses the transpiled code. This has a couple potential benefits:
1. Reducing code duplication. You can see this in a couple PRs, but for example check out the replace step performer before and after. **This PR cuts almost 1k lines of code, while not changing the functionality.**
2. Making us less dependent on Pandas! All we need to do to support another language or platform is to add a new `get_code()` parameter to code chunks.

In general, the `execute` function should pretty much aim to:
1. Update metadata of the step
2. Execute the generated code
3. Potentially compare the result

# Testing

Check out the features. 

# Documentation

No.

### CODE REVIEW

File: mitosheet/mitosheet/step_performers/import_steps/excel_import.py [hunk 1]
```diff
@@ -36,50 +36,46 @@ def step_type(cls) -> str:
...
-        decimal: str = get_param(params, 'decimal')
-
-        read_excel_params = build_read_excel_params(sheet_names, has_headers, skiprows, decimal)
-
-        post_state = prev_state.copy()
-        
         if not os.path.exists(file_name):
             raise make_file_not_found_error(file_name)
+        
+        sheet_names: List[str] = get_param(params, 'sheet_names')
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/import_steps/excel_import.py:42]: There's a functionn that does this?

File: mitosheet/mitosheet/step_performers/import_steps/excel_range_import.py [hunk 2]
```diff
@@ -43,71 +43,43 @@ def step_type(cls) -> str:
...
 
-        if convert_csv_to_xlsx:
-            file_path = convert_csv_file_to_xlsx_file(file_path, sheet_name=sheet['value'])
-
-        post_state = prev_state.copy() 
-
-        pandas_start_time = perf_counter()
-
-        sheet_index_to_df_range: Dict[int, str] = {}
+        new_df_names = []
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/import_steps/excel_range_import.py:48]: Same function. See where else I can use it

File: mitosheet/mitosheet/step_performers/import_steps/snowflake_import.py [hunk 3]
```diff
@@ -67,47 +61,39 @@ def execute(cls, prev_state: State, params: Dict[str, Any]) -> Tuple[State, Opti
...
         
-        try:
-            # Second execute the query
-            cur = con.cursor()
-            sql_query = create_query(table_or_view, query_params)
-            cur.execute(sql_query)
-            df = cur.fetch_pandas_all()
-        except Exception as e: 
-            raise make_invalid_snowflake_import_error(e)
-        finally:
```
Comment (naterush) [inline mitosheet/mitosheet/step_performers/import_steps/snowflake_import.py:78]: Fix up this error!

File: mitosheet/mitosheet/step.py [hunk 4]
```diff
@@ -144,6 +145,8 @@ def set_prev_state_and_execute(self, new_prev_state: State) -> bool:
         execution actually occured.
         """        
         # Saturate the event to get up to date parameters
+        # TODO: this should fill in the execution data - hopefully
+        # we can get all of it without executing. I think we probably can
         params = self.step_performer.saturate(new_prev_state, self.params)
```
Comment (naterush) [inline mitosheet/mitosheet/step.py:150]: Remove the saturate function? I think it's just execute now. I think we want to get rid of it!

**Review by naterush — COMMENTED** [review]
Cleanup

File: mitosheet/mitosheet/step_performers/merge.py [hunk 5]
```diff
@@ -36,146 +37,70 @@ def step_type(cls) -> str:
...
+                params, 
+                execution_data,
+                new_dataframe_params={
+                    'df_source': DATAFRAME_SOURCE_MERGED,
+                    'new_df_names': [new_df_name],
+                    'sheet_indexes': None
+                }
+            )
+
+        except ValueError:
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/merge.py:58]: In this case moving the error handing until after we execute the code seems fine, in other cases we might want to do error handing first. ie: if we know the user is about to launch a super long running operation that we don't want to let them do. But we can still do that by moving it higher in the execute function so this seems fine

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 6]
```diff
@@ -61,18 +67,21 @@ def execute(cls, prev_state: State, params: Dict[str, Any]) -> Tuple[State, Opti
         will allow the logging infrastructure to determine how much overhead Mito
         adds to executing this event.
 
-        If the exectution_data includes the key `result`, than this result can be
-        accessed in the frontend to display some results to the user. See the 
-        DropDuplicatesStepPerformer and the DropDuplicatesCodeChunk for an example of this usage.
+        By default, this function will return the result of just executing the transpiled
+        code. 
+
+        In general, this function should use _no_ Pandas specific functionality, but rather
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/step_performer.py:73]: Currently, we still have some pandas specific code for error handing like in merge btw

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 7]
```diff
@@ -82,6 +91,101 @@ def transpile(
...
+        for chunk in code_chunks:
+            _code, imports = chunk.get_code()
+            code.extend(imports)
+            code.extend(_code)
+
+        final_code = "\n".join(code)
+
+        # NOTE; this is all wrong, but it's just b/c it's hard to get things out of exec. We do our best
+        exec_globals = get_globals_for_exec(post_state, post_state.public_interface_version)
+        exec_locals = {**exec_globals}
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/step_performer.py:131]: I guess this is what your comment is referring to here, but I don't get why we're exceccing the globals and then calling that exec_locals and then later passing both exec_globals and exec_locals to the exec function?

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 8]
```diff
@@ -82,6 +91,101 @@ def transpile(
...
+
+            if len(new_df.columns) > len(old_df.columns):
+                # Find all the new column headers, and add them to the state
+                new_column_headers = [ch for ch in new_df.columns if ch not in old_df]
+                post_state.add_columns_to_state(modified_dataframe_index, new_column_headers)
+
+                # Check if there are any invalid column headers
+                c = Counter(new_df.columns)
+                most_common = c.most_common(1)
+                print(most_common)
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/step_performer.py:158]: remove

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 9]
```diff
@@ -82,6 +91,101 @@ def transpile(
...
+                    column_id = prev_state.column_ids.get_column_id_by_header(modified_dataframe_index, old_header)
+                    post_state.column_ids.set_column_header(modified_dataframe_index, column_id, new_header)
+
+            if len(new_df.columns) > len(old_df.columns):
+                # Find all the new column headers, and add them to the state
+                new_column_headers = [ch for ch in new_df.columns if ch not in old_df]
+                post_state.add_columns_to_state(modified_dataframe_index, new_column_headers)
+
+                # Check if there are any invalid column headers
+                c = Counter(new_df.columns)
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/step_performer.py:156]: Can we turn this column header checking into a utility and add a doc string?

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 10]
```diff
@@ -82,6 +91,101 @@ def transpile(
...
+                sheet_indexes = {}
+
+            for new_df_name in new_dataframe_params['new_df_names']:
+                df_source = new_dataframe_params['df_source']
+                sheet_index = sheet_indexes.get(new_df_name, None)
+
+                new_df = exec_locals[new_df_name] # TODO: this is wrong. This will not always be updated, accoring to exec documentation
+                post_state.add_df_to_state(new_df, df_name=new_df_name, df_source=df_source, sheet_index=sheet_index, use_deprecated_id_algorithm=use_deprecated_id_algorithm)
+
+                print("NEW DF", new_df)
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/step_performer.py:180]: remoe

File: mitosheet/mitosheet/step_performers/transpose.py [hunk 11]
```diff
@@ -31,32 +31,37 @@ def step_type(cls) -> str:
...
         }
 
+        return cls.execute_through_transpile(
+            prev_state, 
+            params, 
+            execution_data, 
+            {
+                'df_source': DATAFRAME_SOURCE_TRANSPOSED,
+                'new_df_names': [new_df_name],
+                'sheet_indexes': None
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/transpose.py:48]: What is the `sheet_indexes` param supposed to be if we are not setting it here?

File: mitosheet/mitosheet/transpiler/transpile_utils.py [hunk 12]
```diff
@@ -349,4 +350,40 @@ def get_default_code_options(analysis_name: str) -> CodeOptions:
         'function_name': 'function_' + analysis_name[-4:], # Give it a random name, just so we don't overwrite them
         'function_params': dict(),
         'import_custom_python_code': False
-    }
\ No newline at end of file
+    }
+
+
+def get_globals_for_exec(state: State, public_interface: int) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (aarondr77) [reply mitosheet/mitosheet/transpiler/transpile_utils.py:356]: todo ^

**Review by aarondr77 — COMMENTED** [review]
This is super cool! Getting rid of so much duplicate code is a big devX win. 

It's working pretty well too!

File: mitosheet/mitosheet/step_performers/step_performer.py (hunk 6 above)
Comment (naterush) [reply mitosheet/mitosheet/step_performers/step_performer.py:79]: Yeah. We can keep refining this over time, but it's all good for now.

File: mitosheet/mitosheet/step_performers/step_performer.py (hunk 9 above)
Comment (naterush) [reply mitosheet/mitosheet/step_performers/step_performer.py:175]: It's simple enough,let's keep it as is for now.

Comment (naterush) [reply mitosheet/mitosheet/step_performers/step_performer.py:175]: Jk, you're right, this function needs cleaning.

**Comment by naterush** [comment]
LGTM once tests pass, and I get another review.

File: mitosheet/mitosheet/code_chunks/postprocessing/set_dataframe_format_code_chunk.py [hunk 13]
```diff
@@ -290,9 +290,16 @@ def get_dataframe_format_code(state: State, sheet_index: int) -> Tuple[Optional[
 class SetDataframeFormatCodeChunk(CodeChunk):
     """
     This is the first postprocessing code chunk. Postprocessing code chunks are executed after all other code chunks,
-    and are given access to the final state. They have no parameters otherwise.
+    and are given access to the final state. This means, unlike other code chunks, it's fine for post-processing code 
+    chunks to get access to the final post_state. 
+    
+    TODO: Maybe they should only get the post state?
```
Comment (marthacryan) [inline mitosheet/mitosheet/code_chunks/postprocessing/set_dataframe_format_code_chunk.py:296]: There could be a parent class `CodeChunk` and children classes like `PostProcessingCodeChunk` and `PreProcessingCodeChunk` that would guide people to do the right thing here

File: mitosheet/mitosheet/step_performers/user_defined_import.py [hunk 14]
```diff
@@ -38,26 +38,29 @@ def step_type(cls) -> str:
...
 
-        # We make a new state to modify it
-        post_state = prev_state.copy()
+        execution_data = {
+            'user_defined_function_params': get_user_defined_function_param_type_and_execute_value_and_transpile_value(prev_state, importer_function, importer_params),
+            'new_df_names': [new_df_name],
+        }
 
-        pandas_start_time = perf_counter()
+        print(execution_data)
```
Comment (marthacryan) [inline mitosheet/mitosheet/step_performers/user_defined_import.py:51]: remove print

File: mitosheet/mitosheet/tests/step_performers/test_melt.py [hunk 15]
```diff
@@ -69,6 +69,10 @@ def test_melt(input_dfs, sheet_index, id_vars, value_vars, output_dfs):
 
     assert len(mito.dfs) == len(output_dfs)
     for actual, expected in zip(mito.dfs, output_dfs):
+        print(actual)
```
Comment (marthacryan) [inline mitosheet/mitosheet/tests/step_performers/test_melt.py:72]: remove print

File: mitosheet/mitosheet/tests/step_performers/test_merge.py [hunk 16]
```diff
@@ -598,15 +598,6 @@ def test_incompatible_merge_key_types_error():
 
     assert mito.transpiled_code == []
 
-
-def test_merge_between_multi_index_and_non_errors():
-    df_one = make_multi_index_header_df({0: [1, 2], 1: [3, 4]}, ['A', ('B', 'count')])
-    df_two = pd.DataFrame({'A': [1, 2], 'B': [5, 6]})
```
Comment (marthacryan) [inline mitosheet/mitosheet/tests/step_performers/test_merge.py:604]: Why are you removing this test?

**Review by marthacryan — COMMENTED** [review]
This is looking amazing! I think this is really a good move to remove a ton of code and make it easier to make changes :D

File: mitosheet/mitosheet/ai/recon.py [hunk 17]
```diff
@@ -200,26 +199,61 @@ def get_modified_dataframe_recon_data(old_df: pd.DataFrame, new_df: pd.DataFrame
...
     }
-    
 
+def delete_column_id_from_state_metadata(
+    state: State,
+    sheet_index: int,
+    column_id: ColumnID,
+) -> State:
+    
+    # And then update all the state variables removing this column from the state
```
Comment (aarondr77) [inline mitosheet/mitosheet/ai/recon.py:267]: Remove "And then"

File: mitosheet/mitosheet/ai/recon.py [hunk 18]
```diff
@@ -200,26 +199,61 @@ def get_modified_dataframe_recon_data(old_df: pd.DataFrame, new_df: pd.DataFrame
...
 
     This includes: handling deleted columns, added columns, renamed columns, and modified columns.
     """
-    old_df = state.dfs[sheet_index]
+    # Check there aren't any duplicated columns in the new dataframe
+    c = Counter(new_df.columns)
+    most_common = c.most_common(1)
+    for ch, count in most_common:
+        if count > 1:
+            raise make_column_exists_error(ch)
```
Comment (aarondr77) [inline mitosheet/mitosheet/ai/recon.py:296]: cool!

File: mitosheet/mitosheet/ai/recon.py [hunk 19]
```diff
@@ -200,26 +199,61 @@ def get_modified_dataframe_recon_data(old_df: pd.DataFrame, new_df: pd.DataFrame
...
 
     # Add new columns to the state
-    state.add_columns_to_state(sheet_index, modified_dataframe_recon['column_recon']['created_columns'])
+    if len(modified_dataframe_recon['column_recon']['created_columns']) > 0:
+        print("CREATED", modified_dataframe_recon['column_recon']['created_columns'], column_headers_to_column_ids)
+        state.add_columns_to_state(
+            sheet_index, 
+            modified_dataframe_recon['column_recon']['created_columns'],
+            column_headers_to_column_ids=column_headers_to_column_ids
+        )
```
Comment (aarondr77) [inline mitosheet/mitosheet/ai/recon.py]: Remove all of these prints

File: mitosheet/mitosheet/step_performers/step_performer.py [hunk 20]
```diff
@@ -153,45 +153,27 @@ def execute_through_transpile(
...
-                most_common = c.most_common(1)
-                for ch, count in most_common:
-                    if count > 1:
-                        raise make_column_exists_error(ch)
-
-        if new_column_params:
-            sheet_index = new_column_params['sheet_index']
-            new_column_headers_to_column_id = new_column_params['new_column_headers_to_column_id']
-            post_state.add_columns_to_state(sheet_index, list(new_column_headers_to_column_id.keys()), new_column_headers_to_column_id)
+            print(new_df)
```
Comment (aarondr77) [inline mitosheet/mitosheet/step_performers/step_performer.py]: remove

File: mitosheet/mitosheet/ai/recon.py [hunk 21]
```diff
@@ -152,8 +207,8 @@ def get_modified_dataframe_recon_data(old_df: pd.DataFrame, new_df: pd.DataFrame
 
     # First, preserving the order, we remove any columns that are in both the old
     # and the new dataframe
-    old_columns_without_shared = list(filter(lambda ch: ch not in new_columns, old_columns))
-    new_columns_without_shared = list(filter(lambda ch: ch not in old_columns, new_columns))
+    old_columns_without_shared = get_added_column_headers(new_columns, old_columns)
```
Comment (aarondr77) [inline mitosheet/mitosheet/ai/recon.py:213]: Maybe a better name for this function is `get_unique_column_headers`

File: mitosheet/mitosheet/tests/step_performers/column_steps/test_change_column_dtype.py [hunk 22]
```diff
@@ -161,6 +159,14 @@ def test_complex_date_strings(strings, result, code):
     else:
         assert len(mito.transpiled_code) == 0
 
+@pandas_post_1_only
```
Comment (aarondr77) [inline mitosheet/mitosheet/tests/step_performers/column_steps/test_change_column_dtype.py:162]: Why did we have to break this test out?

**Review by aarondr77 — COMMENTED** [review]
Nice work!

File: mitosheet/mitosheet/tests/step_performers/test_merge.py (hunk 16 above)
Comment (naterush) [reply mitosheet/mitosheet/tests/step_performers/test_merge.py:604]: We don't actually support multi-index headers currently, this was legacy from when we tried to do so.

**Comment by naterush** [comment]
Merging when tests pass.

**Pull request merged**